### PR TITLE
abstract the storage layer for badger

### PR DESCRIPTION
**Pull request by alonlong**

### CODE REVIEW

**Review by kolesnikovae — COMMENTED** [review]
Looks good. I've left a couple of notes.

In general I'd say that I doubt we need to abstract from a KV storage, YAGNI IMO. Instead, we need an abstraction from the storage at a bit more higher level.

File: pkg/util/timer/timer.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		ticker := time.NewTimer(interval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				// do the function
+				if err := fn(); err != nil {
```
Comment (lzh2nix) [inline pkg/util/timer/timer.go:22]: need handle fn() panic

File: pkg/storage/labels/labels.go [hunk 2]
```diff
@@ -1,93 +1,85 @@
...
-		defer it.Close()
-		for it.Rewind(); it.Valid(); it.Next() {
-			item := it.Item()
-			k := item.Key()
-			shouldContinue := cb(string(k[2:]))
-			if !shouldContinue {
-				return nil
+func (ll *Labels) GetKeys(cb func(k string) bool) error {
+	switch s := ll.db.(type) {
+	case *badger.Service:
```
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/labels/labels.go:36]: Good idea, i will check it later

File: pkg/storage/kv/badger/badger.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func (s *Service) Close() error {
+	if s.done != nil {
+		close(s.done)
+	}
+	if s.db != nil {
+		if err := s.db.Close(); err != nil {
+			logrus.Errorf("close badger: %v", err)
+		}
```
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:137]: ok

File: pkg/storage/kv/badger/badger.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	if err := s.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get([]byte(key))
+		if err != nil {
+			if err == badger.ErrKeyNotFound {
+				return nil
+			}
+			return fmt.Errorf("read from badger: %v", err)
+		}
+		if err := item.Value(func(val []byte) error {
+			data = append([]byte{}, val...)
```
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:98]: ok, will check it

File: pkg/storage/kv/badger/badger.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+	return data, nil
+}
+
+func (s *Service) Set(key, value []byte) error {
+	return s.db.Update(func(txn *badger.Txn) error {
+		if err := txn.Set(key, value); err != nil {
+			return fmt.Errorf("set entry: %v", err)
+		}
+		return nil
```
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:116]: of course, will fix it

File: pkg/util/timer/timer.go (hunk 1 above)
Comment (alonlong) [reply pkg/util/timer/timer.go:22]: Good idea

File: pkg/storage/kv/badger/badger.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:116]: done

File: pkg/storage/kv/badger/badger.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:137]: done

File: pkg/util/timer/timer.go (hunk 1 above)
Comment (alonlong) [reply pkg/util/timer/timer.go:22]: done

File: pkg/storage/kv/badger/badger.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:98]: done

File: pkg/storage/labels/labels.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/labels/labels.go:36]: done

File: pkg/storage/kv/badger/badger.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	db, err := badger.Open(badgerOptions)
+	if err != nil {
+		return nil, fmt.Errorf("badger open: %w", err)
+	}
+
+	// start a timer for the badger GC
+	timer.StartWorker("badger gc", s.done, 5*time.Minute, func() error {
+		logrus.Debug("starting badger garbage collection")
+		return db.RunValueLogGC(0.7)
+	})
```
[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:81]: before close the badger, it will close the timer first

[the remark this answers is not in the document]
Comment (kolesnikovae) [reply pkg/storage/kv/badger/badger.go:81]: The channel will be closed, but there may be ongoing GC - after the channel is closed `Close` should wait for the routine to finish.

[the remark this answers is not in the document]
Comment (alonlong) [reply pkg/storage/kv/badger/badger.go:92]: done

**Review by petethepig — APPROVED** [review]
Looks good

**Comment by petethepig** [comment]
@alonlong Can you fix tests please?

**Comment by alonlong** [comment]
> @alonlong Can you fix tests please?

@petethepig  u mean add unit tests for this PR?

**Pull request closed without merging**